File: app/src/main/java/com/toolbox/tools/core/EngineManager.java
package com.toolbox.tools.core;

import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Objects;

public final class EngineManager {
    private final Map<String, EngineContract> engines = new LinkedHashMap<>();

    public synchronized void register(EngineContract engine) {
        Objects.requireNonNull(engine, "engine");
        if (!engine.isReady()) {
            throw new IllegalArgumentException("Engine is not ready: " + engine.id());
        }
        if (engines.containsKey(engine.id())) {
            throw new IllegalArgumentException("Duplicate engine id: " + engine.id());
        }
        engines.put(engine.id(), engine);
    }

    public synchronized boolean contains(String id) {
        return engines.containsKey(id);
    }

    public synchronized Map<String, EngineContract> snapshot() {
        return Collections.unmodifiableMap(new LinkedHashMap<>(engines));
    }
}

File: app/src/main/java/com/toolbox/tools/core/EngineManager.kt
package com.toolbox.tools.core

class EngineManager {
    private val engines = mutableMapOf<String, EngineContract>()
    private val states = mutableMapOf<String, EngineState>()

    fun addEngine(engine: EngineContract): Boolean {
        if (engine.id.isBlank()) return false
        engines[engine.id] = engine
        states[engine.id] = EngineState.CREATED
        return true
    }

    fun start(id: String): Boolean {
        val engine = engines[id] ?: return false
        engine.start()
        states[id] = EngineState.RUNNING
        return true
    }

    fun stop(id: String): Boolean {
        val engine = engines[id] ?: return false
        engine.stop()
        states[id] = EngineState.STOPPED
        return true
    }

    fun state(id: String): EngineState? = states[id]

    fun hasEngine(id: String): Boolean = engines.containsKey(id)
}
